### PR TITLE
feat: Implement burn-in logic

### PR DESCRIPTION
**Pull request by MuazOthman**
- Closes [CYCLOUD-1665](https://cypress-io.atlassian.net/browse/CYCLOUD-1665)

### Requirements
- [ ] Store the burn-in action state on each mocha runnable instance
- [ ] Refactor the existing experimental retries code
- [ ] Write tests in `system-tests/record_spec.js` for all the enumerated conditions
- [ ] Rewire how actions are stored on `Cypress.__actions`

### PR Tasks

- [ ] Have tests been added/updated?
- [ ] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [ ] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?

[CYCLOUD-1665]: https://cypress-io.atlassian.net/browse/CYCLOUD-1665?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/driver/src/cypress/mocha.ts [hunk 1]
```diff
@@ -490,7 +458,28 @@ function createCalculateTestStatus (Cypress: Cypress.Cypress) {
   Test.prototype.calculateTestStatus = function () {
     const retriesConfig = getNormalizedRetriesConfig(Cypress)
 
-    return calculateTestStatus(this, retriesConfig)
+    //@ts-expect-error
+    const actions = Cypress.__actions
+
+    const action = actions?.find((a) => a.clientId === this.id && a.action === 'BURN_IN') ?? {}
+
+    const latestScore = action?.payload?.startingScore ?? -2
```
Comment (mabela416) [inline packages/driver/src/cypress/mocha.ts]: can we make the change we talked about to?
`const latestScore = action?.payload?.startingScore === undefined ? -2 : action?.payload?.startingScore`

File: packages/app/cypress/e2e/runner/snapshots/runner.mochaEvents.cy.ts.json [hunk 2]
```diff
@@ -1004,7 +1009,8 @@
         "_cypressTestStatusInfo": {
           "shouldAttemptsContinue": false,
           "attempts": 1,
-          "outerStatus": "failed"
+          "outerStatus": "failed",
+          "reasonToStop": "PASSED_FIRST_ATTEMPT"
```
Comment (mabela416) [inline packages/app/cypress/e2e/runner/snapshots/runner.mochaEvents.cy.ts.json]: if the `outerStatus` is failed then how come the reasonToStop is `PASSED_FIRST_ATTEMPT`?

File: packages/driver/src/cypress/mocha.ts [hunk 3]
```diff
@@ -490,7 +458,28 @@ function createCalculateTestStatus (Cypress: Cypress.Cypress) {
...
+
+    // console.log('userFacingBurnInConfig', userFacingBurnInConfig)
+
+    const localBurnInConfig = getBurnInConfig(userFacingBurnInConfig)
+
+    // console.log('localBurnInConfig', localBurnInConfig)
+
+    const completeBurnInConfig = mergeBurnInConfig(action?.payload?.config ?? {}, { values: localBurnInConfig })
+
+    // console.log('completeBurnInConfig', completeBurnInConfig)
```
Comment (chrisbreiding) [inline packages/driver/src/cypress/mocha.ts]: Can these commented-out console logs be removed?

File: packages/driver/src/burnIn/index.ts
Comment (chrisbreiding) [inline packages/driver/src/burnIn/index.ts:1]: Unfortunately, we have some inconsistencies in this repo as far as file name format goes (snake case, pascal case, kebab case), but directory names are pretty consistently kebab-cased. Could we rename this directory to `burn-in`?

File: system-tests/test/record_spec.js [hunk 4]
```diff
@@ -761,6 +761,74 @@ describe('e2e record', () => {
...
+              },
+              action: 'BURN_IN',
+            }],
+          })
+          // }
+
+          // return res.json({
+          //   ...postInstanceTestsResponse,
+          //   actions: [],
+          // })
```
Comment (chrisbreiding) [inline system-tests/test/record_spec.js]: Can the commented code be removed?

File: packages/driver/src/cypress/runner.ts [hunk 5]
```diff
@@ -1597,7 +1597,9 @@ export default {
         )
       },
 
-      run (fn) {
+      run (fn, response) {
+        Cypress.__actions = response?.actions
```
Comment (chrisbreiding) [inline packages/driver/src/cypress/runner.ts:1603]: What is this used for?

**Review by chrisbreiding — CHANGES_REQUESTED** [review]

File: packages/driver/src/cypress/runner.ts (hunk 5 above)
Comment (MuazOthman) [reply packages/driver/src/cypress/runner.ts:1603]: this is used to make the burn-in actions available for the mocha prototype extensions here https://github.com/cypress-io/cypress/blob/15f8fc35c8ba8519a3c7b090b60f0a6f005fe103/packages/driver/src/cypress/mocha.ts#L462

File: packages/driver/src/cypress/mocha.ts (hunk 3 above)
Comment (MuazOthman) [reply packages/driver/src/cypress/mocha.ts]: These comments are now removed

File: packages/driver/src/burnIn/index.ts
Comment (MuazOthman) [reply packages/driver/src/burnIn/index.ts:1]: Fixed

File: system-tests/test/record_spec.js (hunk 4 above)
Comment (MuazOthman) [reply system-tests/test/record_spec.js]: These comments are now removed

File: packages/app/cypress/e2e/runner/snapshots/runner.mochaEvents.cy.ts.json (hunk 2 above)
Comment (MuazOthman) [reply packages/app/cypress/e2e/runner/snapshots/runner.mochaEvents.cy.ts.json]: Thanks for catching this! I fixed it.

**Review by chrisbreiding — APPROVED** [review]

File: packages/driver/src/burn-in/index.ts [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    retriesConfig,
+    maxAttempts,
+    currentAttemptResult,
+    totalAttemptsAlreadyExecuted,
+  } = input
+
+  const result: EvaluateAttemptOutput = { final: true, initialStrategy: potentialInitialStrategy ?? 'NONE' }
+
+  // If there is AT LEAST one failed test attempt, we know we need to apply retry logic.
+  // Otherwise, the test might be burning in (not implemented yet) OR the test passed on the first attempt,
```
Comment (mabela416) [inline packages/driver/src/burn-in/index.ts]: we can remove the parenthesis?

File: packages/driver/src/burn-in/index.ts [hunk 7]
```diff
@@ -0,0 +1,202 @@
...
+          result.reasonToStop = 'FAILED_REACHED_MAX_RETRIES'
+        } else {
+          // we have remaining attempts and we should retry the test
+          result.final = false
+        }
+
+        break
+      default:
+        result.outerTestStatus = 'failed'
+        result.forceState = currentAttemptResult === 'passed' ? currentAttemptResult : undefined
```
Comment (mabela416) [inline packages/driver/src/burn-in/index.ts]: if the `reasonToStop` here is `FAILED_NO_RETRIES` then  `currentAttemptResult` will never be `passed` so will this line always be `undefined`?

**Review by mabela416 — APPROVED** [review]

**Pull request merged**